README.md
# Deprecated
This app is no longer in use at NYPL. Please contact the DXP/RENO team in Digital for more information.

# NYPL Blogs app

| Branch         | Status|  
|:---------------|:---------------------------------------------------------------------------------------------------------------------------|  
| `master`   | [![Build Status](https://travis-ci.org/NYPL/dgx-blogs.svg?branch=master)](https://travis-ci.org/NYPL/dgx-blogs)      |  
| `qa`           | [![Build Status](https://travis-ci.org/NYPL/dgx-blogs.svg?branch=qa)](https://travis-ci.org/NYPL/dgx-blogs) |  
| `development`  | [![Build Status](https://travis-ci.org/NYPL/dgx-blogs.svg?branch=development)](https://travis-ci.org/NYPL/dgx-blogs)      |

## Version
> v0.3.13

## Install

Checkout the `development` branch and perform `nvm use; npm install`

**In OSX, if you encounter install issues**, you may be helped by [this SO post](https://stackoverflow.com/a/52633713/2092409) - which, depending on the state of your Command Line Tools, may just boil down to running:

```
CXXFLAGS="-mmacosx-version-min=10.9" LDFLAGS="-mmacosx-version-min=10.9" npm install
```

## Running Locally

*Note: Running `npm start` or `npm run dist` may dump a bunch of stack traces to stdout. [They may be safe to ignore](#stacktraces-in-logs)*

### Dev Server

To run the app locally, run `npm start` and go to http://localhost:3001/blog/beta/
The Webpack Dev Server will serve static files from localhost:3000

### Production

To run the app in production mode, run:

* npm run dist
* NODE_ENV=production npm start

`npm run dist` will compile and minify the Javascript and SCSS.
`NODE_ENV=production` is an environment variable that tells the app to run in production mode,
and to not start the webpack dev server.

## Contributing

### Git Workflow

 - Cut feature branch from `development`
 - Create PR against `development`
 - After approval, author merges feature into `development` (Note development CD currently disabled)
 - Merge `development` > `qa`
 - Merge `qa` > `master` (production)

## Troubleshooting

### Stacktraces in logs

Under Node v6.11.5 (and v6.10.3), installed dependencies seem to spew a lot of warnings like this:

```
$ npm run dist
> dgx-blogs@0.3.10 dist /Users/_/projects/nypl/dgx-blogs
> NODE_ENV=production webpack --config webpack.config.js

clean-webpack-plugin: /Users/_/projects/nypl/dgx-blogs/dist has been removed.
(node) v8::ObjectTemplate::Set() with non-primitive values is deprecated
(node) and will stop working in the next major release.

==== JS stack trace =========================================

Security context: 0x2f2a10cf781 <JS Object>#0#
    1: .node [module.js:597] [pc=0x1ff6ea41bc44] (this=0x7e9e1695969 <an Object with deprecated map 0x1b1113e1ba11>#1#,module=0x146cb484181 <a Module with map 0x1b1113e1cf61>#2#,filename=0x146cb482859 <String[100]: /Users/_/projects/nypl/dgx-blogs/node_modules/node-sass/vendor/darwin-x64-48/binding.node>)
    2: load [module.js:~478] [pc=0x1ff6ea6aaec3] (this=0x146cb484181 <a Module with map 0x1b1113e1cf61>#2#,filename=0x146cb482859 <String[100]: /Users/_/projects/nypl/dgx-blogs/node_modules/node-sass/vendor/darwin-x64-48/binding.node>)
    3: tryModuleLoad(aka tryModuleLoad) [module.js:446] [pc=0x1ff6ea342afd] (this=0x2f2a1004381 <undefined>,module=0x146cb484181 <a Module with map 0x1b1113e1cf61>#2#,filename=0x146cb482859 <String[100]: /Users/_/projects/nypl/dgx-blogs/node_modules/node-sass/vendor/darwin-x64-48/binding.node>)
    4: _load [module.js:~412] [pc=0x1ff6eaae3e6c] (this=0x7e9e16959e9 <JS Function Module (SharedFunctionInfo 0x7e9e1638c99)>#3#,request=0x146cb481439 <String[100]: /Users/_/projects/nypl/dgx-blogs/node_modules/node-sass/vendor/darwin-x64-48/binding.node>,parent=0x146cb468901 <a Module with map 0x1b1113e1cf61>#4#,isMain=0x2f2a1004271 <false>)
```

Apparently these are [just warnings](https://stackoverflow.com/questions/36897992/nodejs-upgrade-causing-stack-trace) where old packages are hitting deprecation warnings in Node v6. Upgrading a few packages may cause the warnings to disappear.

Note a similar thing is happening in [dgx-new-arrivals](https://github.com/NYPL/dgx-new-arrivals), which seems to share Node version and a lot of boilerplate.

### "error calling API" in logs

At writing, connecting the app to production refinery produces a lot of errors in the logs, which appear to be 404s against this URL:

https://www.nypl.org/refinery/api/v0.1/node/blog?limit=25&offset=1&filter[_enhanced][uri_relative]=blog%2Fbeta%2F

This is also happening in the production EB deployment.

### Front-end bugs

At writing (in OSX 10.14.5, using both FF 67.0.4 & Chrome 75.0.3770.100) following "READ MORE" links from the main blog index (/blob/beta) has a bug where the "LOADING..." overlay never disappears - although blog content is faintly visible beneath.


CHANGELOG.md
## Changelog

### v0.3.14
- Updating @nypl/dgx-react-footer to 0.5.4.

### v0.3.13
- Updating @nypl/dgx-header-component to 2.6.0.

### v0.3.12
- Updating @nypl/dgx-header-component to 2.5.8.

### v0.3.11
- Updating @nypl/dgx-header-component to 2.5.6.

### v0.3.10
- Added fix to skip CrowdStrike installation to prevent deployment failures.

### v0.3.9
- Adding CrowdStrike Falcon sensor to operating system.

### v0.3.8
- Updating @nypl/dgx-react-footer to 0.5.2.

### v0.3.7
- Updating @nypl/dgx-header-component to 2.4.19.

### v0.3.6
- Updating @nypl/dgx-header-component to 2.4.15 and checking for QA in APP_ENV.

### v0.3.5
- Updating @nypl/dgx-header-component to 2.4.14 and setting APP_ENV.

### v0.3.4
- Updating @nypl/dgx-header-component to 2.4.13.

### v0.3.3
- Updating @nypl/dgx-header-component to 2.4.12.

### v0.3.2
- Updating @nypl/dgx-react-footer version to 0.5.1 and @nypl/dgx-header-component to 2.4.11.

### v0.3.1
- Updating @nypl/dgx-react-footer version to 0.5.0 and @nypl/dgx-header-component to 2.4.8.

### v0.3.0
- Updated the Header component to 2.4.7 and added OptinMonster script.

### v0.2.8
- Updated the Header component to v2.4.5.

### v0.2.7
- Updated the Header component to v2.4.2 and Footer component to v0.4.1.

### v0.2.6
- Updated the Header component to v2.4.0.

### v0.2.5
- Updated the Header component to v2.3.0 -- Integrating Fundraising Banner.

### v0.2.4
- Updated the parameters for gaUtils.trackPageview() in App.jsx. It removed unnecessary parameters.

### v0.2.3
- Updated Header to 2.2.0.
- Updated GA initialization configurations.

### v0.2.2
- Updated Header to 2.1.1.

### v0.2.1
- Updated Header to 2.1.0.

### v0.2.0
- Updated to React 15.

### v0.1.12
#### Updated
- Updated the Header Component to v1.5.5. The updates include integrating the log in related functions with login server, removing console loggings for patron token expiration, and turning off the feature flag of OAuth Login and set it as default.

### v0.1.11
#### Updated
- Updated the Header Component to v1.5.1. The update includes HTTPS fix and the JavaScript fallback for the log in button on the Header Component.

### v0.1.10
#### Added
- Enabled Feature Flags plugin on the client-side and added Optimizely script in the index.ejs file.


server.js
import path from 'path';
import express from 'express';
import compress from 'compression';
import colors from 'colors';

import React from 'react';

import { match, RouterContext } from 'react-router';

import ReactDOMServer from 'react-dom/server';

import Iso from 'iso';
import alt from 'dgx-alt-center';

import appConfig from './appConfig.js';
import webpack from 'webpack';
import WebpackDevServer from 'webpack-dev-server';
import webpackConfig from './webpack.config.js';

import appRoutes from './src/app/routes/routes.js';
import apiRoutes from './src/server/ApiRoutes/ApiRoutes.js';

import DocMeta from 'react-doc-meta';

const { appBaseUrl } = appConfig;

const ROOT_PATH = __dirname;
const INDEX_PATH = path.resolve(ROOT_PATH, 'src/client');
const DIST_PATH = path.resolve(ROOT_PATH, 'dist');
const VIEWS_PATH = path.resolve(ROOT_PATH, 'src/views');
const WEBPACK_DEV_PORT = appConfig.webpackDevServerPort || 3000;

const isProduction = process.env.NODE_ENV === 'production';
const app = express();

app.use(compress());

// Disables the Server response from
// displaying Express as the server engine
app.disable('x-powered-by');

// Set the view engine to EJS
app.set('view engine', 'ejs');
app.set('views', VIEWS_PATH);

app.set('port', process.env.PORT || 3001);

// * is used for Reverse Proxy at the moment but can be cleaned up:
// For webpack
app.use('*/dist', express.static(DIST_PATH));
// For images
app.use('*/src/client', express.static(INDEX_PATH));

app.use('/', (req, res, next) => {
  if (req.path === appBaseUrl.slice(0, -1)) {
    return res.redirect(appBaseUrl);
  }
  next();
});

app.use('/', apiRoutes);

app.use('/', (req, res) => {
  const iso = new Iso();
  const blogAppUrl = ((req.url).indexOf('blog') !== -1) && ((req.url).indexOf('blog') < 5);
  const routes = blogAppUrl ? appRoutes.client : appRoutes.server;

  alt.bootstrap(JSON.stringify(res.locals.data || {}));

  match({ routes, location: req.url }, (error, redirectLocation, renderProps) => {
    if (error) {
      res.status(500).send(error.message);
    } else if (redirectLocation) {
      res.redirect(302, redirectLocation.pathname + redirectLocation.search);
    } else if (renderProps) {
      const html = ReactDOMServer.renderToString(<RouterContext {...renderProps} />);

      const metaTags = DocMeta.rewind();
      const renderedTags = metaTags.map((tag, index) =>
        ReactDOMServer.renderToString(<meta data-doc-meta="true" key={index} {...tag} />)
      );

      iso.add(html, alt.flush());
      res
        .status(200)
        .render('index', {
          app: iso.render(),
          appTitle: appConfig.appTitle,
          favicon: appConfig.favIconPath,
          webpackPort: WEBPACK_DEV_PORT,
          appEnv: process.env.APP_ENV,
          path: req.url,
          metatags: renderedTags,
          isProduction,
        });
    } else {
      res.status(404).send('Not found');
    }
  });
});

const server = app.listen(app.get('port'), (error, result) => {
  if (error) {
    console.log(colors.red(error));
  }

  console.log(colors.yellow.underline(appConfig.appName));
  console.log(
    colors.green('Express server is listening at'),
    colors.cyan(`localhost: ${app.get('port')}`)
  );
});

// This function is called when you want the server to die gracefully
// i.e. wait for existing connections
const gracefulShutdown = () => {
  console.log('Received kill signal, shutting down gracefully.');
  server.close(() => {
    console.log('Closed out remaining connections.');
    process.exit(0);
  });
  // if after
  setTimeout(() => {
    console.error('Could not close connections in time, forcefully shutting down');
    process.exit();
  }, 1000);
};
// listen for TERM signal .e.g. kill
process.on('SIGTERM', gracefulShutdown);
// listen for INT signal e.g. Ctrl-C
process.on('SIGINT', gracefulShutdown);


/* Development Environment Configuration
 * -------------------------------------
 * - Using Webpack Dev Server
*/
if (!isProduction) {
  new WebpackDevServer(webpack(webpackConfig), {
    publicPath: webpackConfig.output.publicPath,
    hot: true,
    stats: false,
    historyApiFallback: true,
    headers: {
      'Access-Control-Allow-Origin': 'http://localhost:3001',
      'Access-Control-Allow-Headers': 'X-Requested-With',
    },
  }).listen(3000, 'localhost', (error, result) => {
    if (error) {
      console.log(colors.red(error));
    }
    console.log(colors.magenta('Webpack Dev Server listening at'), colors.cyan('localhost3000'));
  });
}
